src/servers/src/otlp/trace/v1.rs

```diff
@@ -0,0 +1,212 @@
+use opentelemetry_proto::tonic::common::v1::any_value::Value as OtlpValue;
+use pipeline::{GreptimePipelineParams, PipelineWay};
+use session::context::QueryContextRef;
+
+use super::attributes::Attributes;
+use super::span::{parse, TraceSpan};
+use super::{
+    DURATION_NANO_COLUMN, PARENT_SPAN_ID_COLUMN, SPAN_ID_COLUMN, SPAN_KIND_COLUMN,
+    SPAN_NAME_COLUMN, TIMESTAMP_COLUMN, TRACE_ID_COLUMN,
+};
+use crate::error::Result;
+use crate::otlp::utils::{any_value_to_jsonb, make_column_data, make_string_column_data};
+use crate::query_handler::PipelineHandlerRef;
+use crate::row_writer::{self, MultiTableData, TableData};
+
+const APPROXIMATE_COLUMN_COUNT: usize = 30;
+
+/// Convert SpanTraces to GreptimeDB row insert requests.
+/// Returns `InsertRequests` and total number of rows to ingest
+///
+/// Compared with v0, this v1 implementation:
+/// 1. flattens all attribute data into columns.
+/// 2. treat `span_id` and `parent_trace_id` as fields.
+/// 3. removed `service_name` column because it's already in
+///    `resource_attributes.service_name`
+///
+/// For other compound data structures like span_links and span_events here we
+/// are still using `json` data structure.
+pub fn v1_to_grpc_insert_requests(
+    request: ExportTraceServiceRequest,
+    _pipeline: PipelineWay,
+    _pipeline_params: GreptimePipelineParams,
+    table_name: String,
+    _query_ctx: &QueryContextRef,
+    _pipeline_handler: PipelineHandlerRef,
+) -> Result<(RowInsertRequests, usize)> {
+    let spans = parse(request);
+    let mut multi_table_writer = MultiTableData::default();
+
+    let one_table_writer = multi_table_writer.get_or_default_table_data(
+        table_name,
+        APPROXIMATE_COLUMN_COUNT,
+        spans.len(),
+    );
+
+    for span in spans {
+        write_span_to_row(one_table_writer, span)?;
+    }
+
+    Ok(multi_table_writer.into_row_insert_requests())
+}
+
+pub fn write_span_to_row(writer: &mut TableData, span: TraceSpan) -> Result<()> {
+    let mut row = writer.alloc_one_row();
+
+    // write ts
+    row_writer::write_ts_to_nanos(
+        writer,
+        TIMESTAMP_COLUMN,
+        Some(span.start_in_nanosecond as i64),
+        Precision::Nanosecond,
+        &mut row,
+    )?;
+    // write ts fields
+    let fields = vec![
+        make_column_data(
+            "timestamp_end",
+            ColumnDataType::TimestampNanosecond,
+            ValueData::TimestampNanosecondValue(span.end_in_nanosecond as i64),
+        ),
+        make_column_data(
+            DURATION_NANO_COLUMN,
+            ColumnDataType::Uint64,
+            ValueData::U64Value(span.end_in_nanosecond - span.start_in_nanosecond),
+        ),
+    ];
+    row_writer::write_fields(writer, fields.into_iter(), &mut row)?;
+
+    // tags
+    let iter = vec![(TRACE_ID_COLUMN, span.trace_id)]
+        .into_iter()
+        .map(|(col, val)| (col.to_string(), val));
+    row_writer::write_tags(writer, iter, &mut row)?;
```

Comment:
Now that we've set the skipping index for `trace_id`, are there other reasons to set the tag(primary key) for it?
I'm afraid setting `trace_id` for tag may introduce too many lines, which can cause write performance degression.

Reply:
Since 0.12 we will not create inverted index on tag any more so cardinality is not a big issue. In this data modeling, we will still need spans of a particular trace located in sequence physically. This will speed up data retrieval when listing spans of trace.